controllers/vspheremachine_controller.go

```diff
@@ -210,20 +215,89 @@ func (r *VSphereMachineReconciler) reconcileNormal(ctx *context.MachineContext)
+		return reconcile.Result{RequeueAfter: config.DefaultRequeue}, nil
+	}
+
+	// TODO: Replace this with logic to smartly know what service to use
+	var vmService services.VirtualMachineService = &govmomi.VMService{}
 
-	// TODO(akutz) Implement vmService.ReconcileVM
+	// Get or create the VM.
 	vm, err := vmService.ReconcileVM(ctx)
 	if err != nil {
 		return reconcile.Result{}, errors.Wrapf(err, "failed to reconcile VM")
 	}
 
-	// Requeue the operation until the VM is "running".
-	if vmState := vm.State; vmState != infrav1.VirtualMachineStateRunning {
-		ctx.Logger.V(6).Info("requeuing operation until vm state is 'running'", "state", vmState)
+	if vm.State != infrav1.VirtualMachineStateReady {
 		return reconcile.Result{RequeueAfter: config.DefaultRequeue}, nil
 	}
 
+	// reconcile the network information
+	err = reconcileNetwork(ctx, vm, vmService)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// reconcile provider id
+	err = reconcileProviderID(ctx, vm, vmService)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func reconcileNetwork(ctx *context.MachineContext, vm infrav1.VirtualMachine, vmService services.VirtualMachineService) error {
+	expNetCount, actNetCount := len(ctx.VSphereMachine.Spec.Network.Devices), len(vm.Network)
+	if expNetCount != actNetCount {
+		return errors.Errorf("invalid network count for %q: exp=%d act=%d", ctx, expNetCount, actNetCount)
+	}
+	ctx.VSphereMachine.Status.Network = vm.Network
+
+	// Update the MAC addresses in the machine's network config as well.
+	for i := range ctx.VSphereMachine.Spec.Network.Devices {
+		devSpec := &ctx.VSphereMachine.Spec.Network.Devices[i]
+		oldMac, newMac := devSpec.MACAddr, vm.Network[i].MACAddr
+		if oldMac != newMac {
+			devSpec.MACAddr = newMac
+			ctx.Logger.V(6).Info("updating MAC address for device", "network-name", devSpec.NetworkName, "old-mac-addr", oldMac, "new-mac-addr", newMac)
+		}
+	}
+
+	// If the VM is powered on then issue requeues until all of the VM's
+	// networks have IP addresses.
+	var ipAddrs []corev1.NodeAddress
+
+	for _, netStatus := range ctx.VSphereMachine.Status.Network {
+		for _, ip := range netStatus.IPAddrs {
+			ipAddrs = append(ipAddrs, corev1.NodeAddress{
+				Type:    corev1.NodeInternalIP,
+				Address: ip,
+			})
+		}
+	}
+
+	if len(ipAddrs) == 0 {
+		ctx.Logger.V(6).Info("requeuing to wait on IP addresses")
+		return &capiv1errors.RequeueAfterError{RequeueAfter: config.DefaultRequeue}
+	}
+
+	// Use the collected IP addresses to assign the Machine's addresses.
+	ctx.VSphereMachine.Status.Addresses = ipAddrs
+
+	return nil
+}
+
+func reconcileProviderID(ctx *context.MachineContext, vm infrav1.VirtualMachine, vmService services.VirtualMachineService) error {
+	providerID := fmt.Sprintf("vsphere://%s", vm.BiosUUID)
+	if ctx.Machine.Spec.ProviderID == nil || *ctx.Machine.Spec.ProviderID != providerID {
+		ctx.Machine.Spec.ProviderID = &providerID
```

Comment:
Hi, @ykakarap 

Maybe `ProviderID` should be defined in `VSphereMachine` spec ? I guess updating `ctx.Machine.Spec.ProviderID` in the provider controller is invalid, refer to aws provider: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/api/v1alpha2/awsmachine_types.go#L34